src/snowflake/snowpark/mock/_stored_procedure.py

```diff
@@ -0,0 +1,174 @@
+#
+# Copyright (c) 2012-2023 Snowflake Computing Inc. All rights reserved.
+#
+
+import sys
+import typing
+from types import ModuleType
+from typing import Any, Callable, Dict, List, Optional, Tuple, Union
+
+import snowflake.snowpark
+from snowflake.snowpark._internal.udf_utils import (
+    check_python_runtime_version,
+    process_registration_inputs,
+)
+from snowflake.snowpark._internal.utils import TempObjectType
+from snowflake.snowpark.column import Column
+from snowflake.snowpark.exceptions import SnowparkSQLException
+from snowflake.snowpark.mock._plan import calculate_expression
+from snowflake.snowpark.mock._snowflake_data_type import ColumnEmulator
+from snowflake.snowpark.stored_procedure import (
+    StoredProcedure,
+    StoredProcedureRegistration,
+)
+from snowflake.snowpark.types import DataType, _FractionalType, _IntegralType
+
+if sys.version_info <= (3, 9):
+    from typing import Iterable
+else:
+    from collections.abc import Iterable
+
+
+def sproc_types_are_compatible(x, y):
+    if isinstance(x, type(y)):
+        return True
+    elif isinstance(x, _IntegralType) and isinstance(y, _IntegralType):
+        return True
+    elif isinstance(x, _FractionalType) and isinstance(y, _FractionalType):
+        return True
+    return False
+
+
+class MockStoredProcedure(StoredProcedure):
+    def __call__(
+        self,
+        *args: Any,
+        session: Optional["snowflake.snowpark.session.Session"] = None,
+        statement_params: Optional[Dict[str, str]] = None,
+    ) -> Any:
+        args, session = self._validate_call(args, session)
+
+        # Unpack columns if passed
+        parsed_args = []
+        for arg, expected_type in zip(args, self._input_types):
+            if isinstance(arg, Column):
+                expr = arg._expression
+                if not sproc_types_are_compatible(expr.datatype, expected_type):
+                    raise ValueError(
+                        f"Unexpected type {expr.datatype} for sproc argument of type {expected_type}"
+                    )
+                parsed_args.append(
+                    calculate_expression(
+                        expr,
+                        ColumnEmulator(data=[None]),
+                        session._analyzer,
+                        {},
+                        keep_literal=True,
```

Comment:
I have two dumb questions here:

1. what is `Column` type args for sproc and how this is used by users?
2. why we set `ColumnEmulator(data=[None]),` here, this gives us an empty column with just the name, is this intended? this is kinda relavant to the first question on how `Column` is used in sproc.

Reply:
1. Some of our tests pass in `lit` values rather than raw values. Those are the only Column objects I've observed being passed to sproc.
2. The way calculate_expression works for lit values requires there to be some input data even if it is all null. I think I can refactor this so that it is more clear.

Reply:
These are good questions actually. In fact I feel like we shouldn't support column inputs to stored procedures, although that might be a limitation from `DataFrame.select`...

Reply:
+1 for refactor to be more clear